lib/derivations.nix

```
@@ -64,18 +77,44 @@ in
       #
       # This can be used for adding package attributes, such as `tests`.
       passthru ? { }
+    , # Optional list of assumed outputs. Default: ["out"]
+      #
+      # This must match the set of outputs that the returned derivation has.
+      # You must use this when the derivation has multiple outputs.
+      outputs ? [ "out" ]
     }:
     let
       # These checks are strict in `drv` and some `drv` attributes, but the
       # attrset spine returned by lazyDerivation does not depend on it.
       # Instead, the individual derivation attributes do depend on it.
       checked =
         throwIfNot (derivation.type or null == "derivation")
-          "lazySimpleDerivation: input must be a derivation."
+          "lazyDerivation: input must be a derivation."
           throwIfNot
-          (derivation.outputs == [ "out" ])
-          # Supporting multiple outputs should be a matter of inheriting more attrs.
-          "The derivation ${derivation.name or "<unknown>"} has multiple outputs. This is not supported by lazySimpleDerivation yet. Support could be added, and be useful as long as the set of outputs is known in advance, without evaluating the actual derivation."
+          (derivation.outputs == outputs)
```

Oh actually, how about
```suggestion
          (all (output: elem output derivation.outputs) outputs)
```

This makes sure that _at least_ all assumed outputs actually exist, without requiring all actual outputs to be given.

---

That would work in a technical sense, but consider the case where an output is added in the underlying derivation, such as `dev`. `lazyDerivation` would remove it and cause it to fail as a `buildInputs` item, without any indication as to what happened.

Now I don't think `lazyDerivation` should be used on arbitrary derivations anyway, for reasons like that, but that also means that - usually - users should be able to know the correct list of outputs and have no problem with this restriction.

---

I've added an explanatory comment in the code.